File: temoto_nl_interface/actions/find_coordinates/src/get_coordinates_scripts/getcoord_origincoord_return.cpp
#include <opencv2/opencv.hpp>
#include <nlohmann/json.hpp>
#include <stdexcept>
#include <string>

using json = nlohmann::json;

class CoordinateConverter {
public:
    static std::string get(const std::string& assistant_reply, 
                            double resolution, 
                            const std::pair<double, double>& origin, 
                            const cv::Size& map_shape, 
                            double angle) {
        try {
            // Parse the assistant reply
            json assistant_reply_dict = json::parse(assistant_reply);

            // Extract coordinates
            json& coords = assistant_reply_dict["coordinates"];
            
            // Check if coordinates exist
            if (!coords.contains("x") || !coords.contains("y")) {
                throw std::runtime_error("Invalid coordinates in assistant reply");
            }

            // Extract image coordinates
            double x_img = coords["x"];
            double y_img = coords["y"];

            // Get origin coordinates
            double origin_x = origin.first;
            double origin_y = origin.second;

            // Map dimensions
            int map_height = map_shape.height;

            // Convert image coordinates to world coordinates
            double x_world = x_img * resolution + origin_x;
            double y_world = (map_height - y_img) * resolution + origin_y;

            // Update coordinates in the dictionary
            coords["x"] = x_world;
            coords["y"] = y_world;

            // Add angle to the response
            assistant_reply_dict["angle"] = angle;

            // Return the updated JSON string
            return assistant_reply_dict.dump();

        } catch (const std::exception& e) {
            // Create error response
            json error_response = {
                {"success", "false"},
                {"error", "ConversionError"},
                {"message", std::string("Error converting coordinates: ") + e.what()}
            };

            return error_response.dump();
        }
    }
};

File: temoto_nl_interface/actions/find_coordinates/src/getcoord_py/getcoord_origincoord_return.py
# getcoord_origincoord_return.py

import cv2
import numpy as np
import yaml
import random
import math
import json


def get(assistant_reply, resolution, origin, map_shape, angle):
    """
    Converts coordinates from the image coordinate system (origin at top-left) to real-world coordinates.

    Parameters:
        assistant_reply (str): A JSON-formatted string containing coordinates in the image coordinate system.

    Returns:
        str: A JSON-formatted string with coordinates converted to the real-world coordinate system.
    """

    try:
        map_height, map_width = map_shape

        # Parse assistant_reply
        assistant_reply_dict = json.loads(assistant_reply)
        coords = assistant_reply_dict.get('coordinates', {})
        x_img = coords.get('x')
        y_img = coords.get('y')

        if x_img is None or y_img is None:
            # Handle error
            error_message = "Invalid coordinates in assistant reply."
            return json.dumps({
                "success": "false",
                "error": "InvalidCoordinates",
                "message": error_message
            })

        # Get origin and resolution
        origin_x, origin_y = origin[0], origin[1]
        resolution = resolution

        # Convert image coordinates to real-world coordinates
        x_world = x_img * resolution + origin_x
        y_world = (map_height - y_img) * resolution + origin_y  # Adjusted line

        # Update the assistant_reply_dict with world coordinates
        assistant_reply_dict['coordinates']['x'] = x_world
        assistant_reply_dict['coordinates']['y'] = y_world

        # Include angle to response
        assistant_reply_dict['angle'] = angle

        assistant_reply_updated = json.dumps(assistant_reply_dict)

        # Return the updated response as JSON
        return assistant_reply_updated  # Removed unnecessary json.dumps

    except Exception as e:
        error_message = f"Error converting coordinates: {e}"
        return json.dumps({
            "success": "false",
            "error": "ConversionError",
            "message": error_message
        })
